MTLEndModel training: save model with best validation score on a specific task

astarostap opened the pull request:
Add a parameter "t" to the MTLEndModel score method so user can specify which task they would like to score.

Also added a "validation_task" param to the Classifier.train_model method, analogous to the existing "validation_metric" param. If "validation_task" is passed then the checkpoint will keep the model that scores the highest for that task's validation_metric. Otherwise it will work as before.

This shouldn't affect the single EndModel.

ajratner commented:
@astarostap Could you get this to pass tests first, then ping us again (just tag us with an @ mention here)?  Thanks!

ajratner commented:
Actually, I think this might be the same `make check` issue that we were having elsewhere... @bhancock8 any luck figuring out what's going on here?

astarostap commented:
@ajratner yeah I'm not sure what's going on, seems to be failing on `make check` but it's passing locally.

astarostap commented:
@ajratner @bhancock8 passing tests. The issue was that I forgot to rebase before pushing the branch.

ajratner commented on metal/classifier.py:
```diff
@@ -473,7 +474,7 @@ def predict(self, X, break_ties="random", return_probs=False, **kwargs):
             [Optionally: Y_s: An [n, k] np.ndarray of predicted probabilities]
         """
-        Y_s = self._to_numpy(self.predict_proba(X, **kwargs))
+        Y_s = self._to_numpy(self.predict_proba(X))
```
Presumably this is due to passing the `t` kwarg into the `Classifier.score` function above; see suggested fix

ajratner commented on metal/classifier.py:
```diff
@@ -236,6 +236,7 @@ def _train_model(self, train_data, loss_fn, dev_data=None, log_writer=None):
                     metric=val_metric,
                     verbose=False,
                     print_confusion_matrix=False,
+                    t=train_config.get("validation_task"),
```
Rather than passing a specific kwarg in in the base class `Classifier`- which could cause issues as below- I think should just create a new dictionary in the config file, like "validation_scoring_kwargs", and then pass this whole dictionary in as the kwargs here (e.g. `**validation_scoring_kwargs`)?  Then, in the single-task end model, this can just be defined as an empty dict... and here, this can contain `t: ...`

astarostap replied:
Got it. I'm trying to do that but I'm running into an issue. I want the argument to be called "validation_task" for the train_model method, and be called "t" for the MTEndModel score method. 

That way for example you'd call mt_em.train_model(validation_task=0..) and mt_em.score(t=0...)

Do you agree that naming makes sense? do you've an idea of how to do that if we go this route?

What I'm able to do atm is to call it either "t" or "validation_task" for both methods

astarostap replied:
Sounds good 👍

ajratner commented on metal/multitask/mt_em_defaults.py:
```diff
@@ -6,4 +6,5 @@
     #   [list]: specify explicitly the layer for each head
     "pass_predictions": False,
     # If True, pass output of parent tasks as additional input to children tasks
+    "train_config": {"validation_task": None},
```
See suggested change above

astarostap commented:
@ajratner I changed the param to the format you said, within the train_configs. Wasn't sure if you meant within tran_configs or outside.

Also I fixed a couple bugs in mt_classifier.score_task that I found writing the unit tests. I think that method wasn't being used anywhere and didn't have unit tests. We might reconsider taking that method out since now the normal score function also works for scoring a specific task.

bhancock8 commented:
lgtm! And thanks for adding the test for score_task(). You get a good citizen gold star. :)

astarostap commented:
@bhancock8 yay gold star :p

@ajratner waiting for your lgtm

ajratner commented:
Awesome!  I think for now we can keep the separate `score_task` function.  Per the class docstrings, the idea is that these `*_task` methods all end up calling `predict_task_proba`, which could be implemented to be faster than predicting all the tasks.  This may be a bit overkill, so we may indeed want to remove all of these?  But can leave for another PR / refactor.  Anyway, LGTM, merging in!
